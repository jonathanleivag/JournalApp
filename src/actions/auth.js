import { types } from "../types/types";
import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import { finishLoading, startLoading } from "./ui";
import Swal from "sweetalert2";

export const startRegister = (name, email, password) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        Swal.fire("Error", error.message, "error");
      });
  };
};

export const startLogin = (email, password) => {
  return async (dispatch) => {
    dispatch(startLoading());
    try {
      const {
        user: { uid, displayName },
      } = await firebase.auth().signInWithEmailAndPassword(email, password);
      dispatch(login(uid, displayName));
      dispatch(finishLoading());
    } catch (error) {
      dispatch(finishLoading());
      Swal.fire("Error", error.message, "error");
    }
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user: { uid, displayName } }) => {
        dispatch(login(uid, displayName));
      })
      .catch((error) => Swal.fire("Error", error.message, "error"));
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: { uid, displayName },
});

export const startLogout = () => {
  return async (dispatch) => {
    try {
      await firebase.auth().signOut();
      dispatch(logout());
    } catch (error) {
      Swal.fire("Error", error.message, "error");
    }
  };
};

export const logout = () => ({ type: types.logout });
