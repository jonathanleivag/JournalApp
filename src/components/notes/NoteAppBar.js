import React, { useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploading } from "../../actions/notes";

// react-journal

// https://api.cloudinary.com/v1_1/dmjxpnr5g

export const NoteAppBar = () => {
  const { active } = useSelector((state) => state.notes);
  const dispatch = useDispatch();
  const handlerSave = () => {
    dispatch(startSaveNote(active));
  };

  const style = { display: "none" };

  const ref = useRef();

  const handlerPictureClick = () => ref.current.click();

  const handlerChangePicture = () => {
    const file = ref.current.files[0];
    if (file) dispatch(startUploading(file));
  };

  return (
    <div className="notes__appbar">
      <span>28 de Agosto 2020</span>

      <input
        type="file"
        id="file"
        style={style}
        onChange={handlerChangePicture}
        ref={ref}
        accept="image/*"
      />

      <div>
        <button className="btn" onClick={handlerPictureClick}>
          Picture
        </button>
        <button className="btn" onClick={handlerSave}>
          Save
        </button>
      </div>
    </div>
  );
};
