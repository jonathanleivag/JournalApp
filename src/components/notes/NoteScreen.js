import React, { useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNote, startDelete } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NoteAppBar } from "./NoteAppBar";

export const NoteScreen = () => {
  const { active: note } = useSelector((state) => state.notes);

  const [values, handleInputChange, reset] = useForm(note);
  const { title, body } = values;

  const { url } = note;

  const refNoteId = useRef(note.id);

  const dispatch = useDispatch();

  useEffect(() => {
    if (refNoteId.current !== note.id) {
      reset(note);
      refNoteId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNote(values.id, { ...values }));
  }, [values, dispatch]);

  const handlerDelete = () => {
    dispatch(startDelete(refNoteId));
  };

  return (
    <div className="notes__main-content">
      <NoteAppBar />
      <div className="notes__content">
        <input
          type="text"
          placeholder="Some awesome title"
          className="notes__title-input"
          name="title"
          autoComplete="off"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          placeholder="What happened today"
          className="notes__textarea"
          autoComplete="off"
          name="body"
          value={body}
          onChange={handleInputChange}
        ></textarea>

        {url && (
          <div className="notes__image">
            <img src={url} alt="Imagen" />
          </div>
        )}
      </div>
      <button className="btn btn-danger" onClick={handlerDelete}>
        Delete
      </button>
    </div>
  );
};
