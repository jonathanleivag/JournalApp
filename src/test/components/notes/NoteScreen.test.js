import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { NoteScreen } from "../../../components/notes/NoteScreen";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { activeNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {
    uid: "epzyh6acV6bvzK60qwGony1xC0t2",
    name: "Johnny Leiva GÃ³mez",
  },
  ui: {
    loading: false,
    msgError: null,
  },
  notes: {
    notes: [
      {
        id: "8Xc6u3bqWoplmT5a15ts",
        title: "",
        body: "",
        date: 1601788561952,
      },
    ],
    active: {
      id: "8Xc6u3bqWoplmT5a15ts",
      date: 1601788561952,
      title: "Hola que tal",
      body: "hola que tal",
    },
  },
};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <NoteScreen />
  </Provider>
);

jest.mock("../../../actions/notes", () => ({
  activeNote: jest.fn(),
}));

describe("Pruebas en <NoteScreen />", () => {
  test("debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de disparar el active note", () => {
    wrapper.find('input[name="title"]').simulate("change", {
      target: {
        name: "title",
        value: "hola que tal",
      },
    });

    expect(activeNote).toHaveBeenLastCalledWith('8Xc6u3bqWoplmT5a15ts', {
      body: "hola que tal",
      title: "hola que tal",
      id: '8Xc6u3bqWoplmT5a15ts',
      date: 1601788561952,
    });
  });
});
