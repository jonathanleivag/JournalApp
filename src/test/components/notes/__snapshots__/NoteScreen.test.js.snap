// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de Agosto 2020
    </span>
    <input
      accept="image/*"
      id="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Hola que tal"
    />
    <textarea
      autoComplete="off"
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="hola que tal"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
